ably/event_emitter.go

```diff
@@ -0,0 +1,176 @@
+package ably
+
+import (
+	"runtime/debug"
+	"sync"
+)
+
+type eventEmitter struct {
+	sync.Mutex
+	listeners listenersForEvent
+	logger    Logger
+}
+
+type emitterEvent = interface{}
+type emitterData = interface{}
+
+type listenersForEvent map[emitterEvent]listenerSet
+type listenerSet map[*eventListener]struct{}
+
+type eventListener struct {
+	handler func(emitterData)
+	once    bool
+
+	queueMtx sync.Mutex
+	queue    []emitterData
+}
+
+func (l *eventListener) handle(e emitterData, log Logger) {
+	// The goroutine that finds the queue empty launches a goroutine for emptying it,
+	// ie. process its emitterData and afterwards any emitterData that concurrent
+	// goroutines may have left in the queue in the meantime.
+	//
+	// Other goroutines just enqueue their emitterData.
+	//
+	// This ensures our guarantee that a single handler runs concurrently with
+	// other handlers, but calls to the same handler are sequential and ordered,
+	// while emitting an event doesn't block the emitting goroutine.
+
+	var isBusy bool
+
+	l.queueMtx.Lock()
+	isBusy = len(l.queue) > 0
+	l.queue = append(l.queue, e)
+	l.queueMtx.Unlock()
+
+	if isBusy {
+		return
+	}
+
+	go func() {
+		done := false
+		for !done {
+			l.queueMtx.Lock()
+			e := l.queue[0]
+			l.queueMtx.Unlock()
+
+			safeHandle(e, l.handler, log)
+
+			l.queueMtx.Lock()
+			l.queue = l.queue[1:]
+			done = len(l.queue) == 0
+			l.queueMtx.Unlock()
+		}
+	}()
+}
+
+func safeHandle(e emitterData, handle func(emitterData), log Logger) {
+	defer func() {
+		r := recover()
+		if r == nil {
+			return
+		}
+		log.Printf(LogError, "EventEmitter: panic in event handler: %v\n%s", r, debug.Stack())
+	}()
+
+	handle(e)
+}
+
+func newEventEmitter(logger Logger) *eventEmitter {
+	return &eventEmitter{
+		listeners: listenersForEvent{
+			nil: listenerSet{},
+		},
+		logger: logger,
+	}
+}
+
+// On registers an event listener. If event is nil, all events will trigger
```

Comment:
I decided to use nil in order to avoid deviating too much from the spec.

Reply:
I don't have a strong opinion about what consideration is more important, though.

Reply:
For me, allowing `nil` to express a desire for 'no filtering' in this case feels natural, especially if that matches what's in the spec.

Having said that, and given my lack of any experience in Go programming, `nil` [sounds complicated](https://go101.org/article/nil.html) so I may well be carrying assumptions with me from other languages.

Reply:
Ah ok, so I definitely had my Go hat on here where I always try to avoid overloading the semantics of `nil` to mean something other than the zero value of a pointer.

Looking at the spec, I see it declares that `On` should support two method signatures:

```
class EventEmitter<Event, Data>:
  on((Data...) ->) // RTE4
  on(Event, (Data...) ->) // RTE4
```

Given this spec covers so many different languages, I think it's incorrect to assume every language supports method overloading, but anyway, for Go I think we should implement two different methods rather than trying to represent overloading with `nil`:

```go
func (e *EventEmitter) On(func(data)) { }
func (e *EventEmitter) OnEvent(event, func(data)) { }
```

If you disagree, then keep the `nil` 🙂

Reply:
That's a good point. Translating overloads from the spec as different methods rather than by adopting a "lowest common denominator" kind of thing with nils sounds good to me.

Done at f6a700e.